tunnel-obfuscation/src/shadowsocks.rs

```diff
@@ -0,0 +1,258 @@
+            server,
+            _shutdown_tx: shutdown_tx,
+        })
+    }
+}
+
+async fn run_obfuscation(
+    local_udp_socket: UdpSocket,
+    shadowsocks_endpoint: SocketAddr,
+    wireguard_endpoint: SocketAddr,
+    shutdown_rx: oneshot::Receiver<()>,
+    #[cfg(target_os = "linux")] fwmark: Option<u32>,
+) -> Result<()> {
+    wait_for_local_udp_client(&local_udp_socket)
+        .await
+        .map_err(Error::WaitForUdpClient)?;
+
+    let shadowsocks = create_shadowsocks_client(
+        shadowsocks_endpoint,
+        #[cfg(target_os = "linux")]
+        fwmark,
+    )
+    .await
+    .map_err(Error::ConnectShadowsocks)?;
+
+    let local_udp = Arc::new(local_udp_socket);
+    let shadowsocks = Arc::new(shadowsocks);
+
+    let wg_addr = Address::SocketAddress(wireguard_endpoint);
+
+    let mut client = tokio::spawn(handle_outgoing(
+        shadowsocks.clone(),
+        local_udp.clone(),
+        wg_addr.clone(),
+    ));
+    let mut server = tokio::spawn(handle_incoming(shadowsocks, local_udp, wg_addr));
+
+    tokio::select! {
+        _ = shutdown_rx => {
+            log::trace!("Stopping shadowsocks obfuscation");
+        }
+        _result = &mut server => log::trace!("Shadowsocks client closed"),
+        _result = &mut client => log::trace!("Local UDP client closed"),
+    }
+
+    client.abort();
+    let _ = client.await;
+    server.abort();
+    let _ = server.await;
+
+    Ok(())
+}
+
+async fn create_shadowsocks_client(
+    shadowsocks_endpoint: SocketAddr,
+    #[cfg(target_os = "linux")] fwmark: Option<u32>,
+) -> std::result::Result<ProxySocket, ProxySocketError> {
+    let ss_context = Context::new_shared(ServerType::Local);
+    let ss_config: ServerConfig = ServerConfig::new(
+        shadowsocks_endpoint,
+        SHADOWSOCKS_PASSWORD,
+        SHADOWSOCKS_CIPHER,
+    );
+    let connect_opts = ConnectOpts {
+        #[cfg(target_os = "linux")]
+        fwmark,
+        ..Default::default()
+    };
+    ProxySocket::connect_with_opts(ss_context, &ss_config, &connect_opts).await
+}
+
+async fn create_local_udp_socket(ipv4: bool) -> Result<(UdpSocket, SocketAddr)> {
+    let random_bind_addr = if ipv4 {
+        SocketAddr::new("127.0.0.1".parse().unwrap(), 0)
+    } else {
+        SocketAddr::new("::1".parse().unwrap(), 0)
+    };
+    let local_udp_socket = UdpSocket::bind(random_bind_addr)
+        .await
+        .map_err(Error::BindUdp)?;
+    let udp_client_addr = local_udp_socket
+        .local_addr()
+        .map_err(Error::GetUdpLocalAddress)?;
+
+    Ok((local_udp_socket, udp_client_addr))
+}
+
+/// Wait for a client to connect to `udp_listener` and connect the socket to that address
+async fn wait_for_local_udp_client(udp_listener: &UdpSocket) -> io::Result<()> {
+    let mut client_rx_buffer = vec![0u8; u16::MAX as usize];
+
+    log::trace!("Waiting for UDP socket client");
+    let (_bytes_n, client_addr) = udp_listener.peek_from(&mut client_rx_buffer).await?;
```

Comment:
Could `UdpSocket::peek_sender` be simpler solution?

Reply:
Cool. I didn't know about that one. Fixed!